### PR TITLE
Cut a few more queries over to registerQuery

### PR DESCRIPTION
**Pull request by nik9000**
Basically just copy-and-paste the fromXContent method and remove the old class.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/index/percolator/PercolatorQueryCacheTests.java [hunk 1]
```diff
@@ -105,9 +103,12 @@ void initialize(Object... fields) throws IOException {
                 .build();
 
         Map<String, Tuple<ParseField, QueryParser<?>>> queryParsers = new HashMap<>();
-        queryParsers.put(TermQueryBuilder.NAME, new Tuple<>(TermQueryParser.QUERY_NAME_FIELD, new TermQueryParser()));
+        // It isn't clear to me why the casts are needed but ok
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/percolator/PercolatorQueryCacheTests.java]: this is funny indeed, the following in two lines doesn't require casting:
```
QueryParser<TermQueryBuilder> queryParser = TermQueryBuilder::fromXContent;
        queryParsers.put(TermQueryBuilder.NAME, new Tuple<>(TermQueryBuilder.QUERY_NAME_FIELD, queryParser));
```

settle for one of the variants and remove the comment?

File: core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java [hunk 2]
```diff
@@ -335,4 +337,53 @@ public void testNonDefaultSimilarity() throws Exception {
...
+     * Failure should not change (and the value should never match anything...).
+     */
+    public void testThatUnrecognizedFromStringThrowsException() {
+        try {
+            HasChildQueryBuilder.parseScoreMode("unrecognized value");
+            fail("Expected IllegalArgumentException");
+        } catch (IllegalArgumentException e) {
+            assertThat(e.getMessage(), is("No score mode for child query [unrecognized value] found"));
+        }
+    }
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java]: we could potentially have a single test for all these above with randomized values?

**Comment by javanna** [comment]
pretty much same as https://github.com/elastic/elasticsearch/pull/17552#issuecomment-206236303 LGTM based on trust besides two minor comments on testing

File: core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java (hunk 2 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java]: I wanted to copy and paste as much as possible here. I can certainly combine if you'd like.

File: core/src/test/java/org/elasticsearch/index/percolator/PercolatorQueryCacheTests.java (hunk 1 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/index/percolator/PercolatorQueryCacheTests.java]: Ooh! I'll see about that.

File: core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java (hunk 2 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java]: oh so this is not new stuff? If it was already there this way leave it, please!

Comment (nik9000) [inline core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java]: I didn't want to change these tests to keep the "I just copied and pasted and didn't make extensive changes" nature of the change.

Comment (javanna) [inline core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java]: ++ thanks for doing it this way, I had thought it was new stuff in the beginning. looks good as is.

**Comment by nik9000** [comment]
Merged! Thanks for the tip on the testing @javanna ! I see light at the end of this tunnel.

**Pull request merged**